replace transit with protobuf

cammellos opened the pull request:
Updates spec to protobuf.

cammellos committed:
replace transit with protobuf

decanus commented on status-group-chats-spec.md:
```diff
@@ -10,7 +10,7 @@
 - [Abstract](#abstract)
 - [Membership updates](#membership-updates)
-  - [Chat ID](#chat-id)
+  - [Chat ID](#chatId)
```
should be `#chat-id` so links work

decanus commented on status-group-chats-spec.md:
````diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
-The format is:
+The protobuf messages are:
 
 ```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
+The protobuf description is:
 
-<!-- TODO: Be more precise in struct description, a la RFC, e.g. TLS style https://tools.ietf.org/html/rfc8446 -->
+```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
 
-Contact requests consists in 3 kind of messages: `ContactRequest`, `ContactRequestConfirmed` and `ContactUpdate`.
-These messages are used to notify the receiving end that it has been added to the sender's contact. They are identified by the transit tags `c2`, `c3`, `c4` respectively, but they are all interchangeable, meaning a client SHOULD handle them in exactly the same way.  The payload of the 3 messages is identical.
+```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
+
+`SyncInstallationContact` messages are used to synchronize in a best-effort the contacts to other devices.
+
+```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
+A `ChatMessage` with `STICKER` `Content/Type` MUST also specify the ID of the `Pack` and 
+the `Hash` of the pack, in the `Sticker` field of `ChatMessage`
+
+```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
-`SyncInstallation` messages are used to synchronize in a best-effort way all the paired installations. It is identified by a transit tag of `p1` 
+`SyncInstallationPublicChat` message is used to synchronize in a best-effort the public chats to other devices.
+
+```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
-`PairInstallation` messages are used to propagate informations about a device to its paired devices. It is identified by a transit tag of `p2` 
+`PairInstallation` messages are used to propagate informations about a device to its paired devices.
+
+```
````
```suggestion
```protobuf
```

decanus commented on status-payloads-spec.md:
````diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
+`MembershipUpdateEvent` is a message used to propagate information about group membership changes in a group chat.
 The details are in the  [Group chats specs](status-group-chats-spec.md)
 
+```
````
```suggestion
```protobuf
```

arnetheduck commented on status-group-chats-spec.md:
````diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
-```
-
-This structure is then stringified collapsing all whitespaces and the `Keccak256` of the string is then signed using its private key by the author and added to the payload.
+The signature for each event is calculated by encoding each `MembershipUpdateEvent` in its protobuf representation
````
protobuf is a non-deterministic format - there are many valid serializations for the same data - this will not fly.

arnetheduck replied:
fwiw,  here's the relevant documentation link: https://developers.google.com/protocol-buffers/docs/encoding#order

cammellos replied:
I don't think that's a problem, I believe we had already a discussion about this at some point.

Basically the fact that protobuf does not guarantees a deterministic format (across versions I believe, same implementation generally is deterministic, but let's assume as well it isn't), does not impact this, for example:

1) I create an event, I encode it to protobuf and sign the bytes generated, send over `signature+bytes`.
2) A client receives it, verify `signature` against `bytes`, decodes the message. 

At no point a client should re-encode the message nor the message should be re-encoded (as well because that's likely to break in protobuf, as unknown fields are simply ignored, so it would break the protocol at the first upgrade), if the message is to be relayed, the original `signature+bytes` should be used, unchanged.

The event itself when it comes to the signature, is treated as just as bytes, the fact that generating those bytes is not deterministic as no impact on the validity of the signature.

If client A (for whatever reason, even if currently there's no use case for this, as re-transmitting uses the raw payload), re-encodes the message twice, and gets two different results, it is also not a problem as the protocol is idempotent so no change will result from duplicates, even if they are not recognized as such.

What do you think?

arnetheduck replied:
I think this is fishy because it becomes possible to sign byte sequences whose deserialized representation is unexpected - you're signing some random bytes at that point and not a particular message as it logically appears to the receiver - you're proving that you have the private key, but not that it was your intent to sign a particular instruction, which seems to open up the signature for interpretation as well, weakening the value of it.

I agree it's mitigated by the custom (non-protobuf) serialization format in `events` and that the use case is somewhat casual (chat). I anticipate surprises as it starts getting abused.

> unknown fields are simply ignored

not sure what you mean here - it's a feature of protobuf to be able to reencode and alter a message without understanding all the fields - that includes preserving unknown fields (perhhaps not all protobuf implementations do this currently, c++/python did last I checked)

decanus replied:
What we did in MVDS is order the fields to sign, see `message indentifiers`. https://specs.vac.dev/mvds.html#payloads

cammellos replied:
Not sure I fully understand your point, 
you are not signing some "random bytes", you are signing the encode version of the message you are about to send.
It does imply that you trust the protobuf library to encode the message non maliciously, is that your concern? (that goes to any library that you use to generate the message though, whether you are concatenating the field etc, regardless of the signature scheme)

The method used is the one suggested by the primary author of protobuf https://groups.google.com/d/msg/protobuf/iCA5ALRtDUI/vUiAl6GQqUAJ &  https://capnproto.org/ .

The alternative so far proposed (to sign each field, concatenating etc), are fine with static protocol that don't change often, that's why we used it in the first place in this one, but are more prone to break as newly added field might be ignored by older clients, so won't be used to calculate the signature.

Unkowns fields were ignored in proto 3, seems like they corrected it in proto 3.5 (we still use 3 ourselves):

https://developers.google.com/protocol-buffers/docs/proto3#unknowns

So the method used before (concatenating each field separately), would break when adding a new field (or we have to have a new field with the new signature scheme, maintaining the old one).

What method are you proposing exactly as an alternative?

MVDS strategy for example, given we use protobuf3 , would break if we ever added a new field to the datasync protobuf message, and the choice would be, either to exclude it from messageID calculation (which might be fine with messageID maybe, definitely not with signatures), or account that with backward compatible code.

arnetheduck replied:
Assuming that the reading client parses the data and then displays it to the user as "signed", there could be question about what exactly was signed: was it the parsed message or the byte-stream? I wonder then if the sender can construct a message, sign it and then claim that what they signed was not what the recipient understood it to be (after parsing).

You make a good point that if idempotency is not needed, the serialization matters less - in MVDS the message id is important for the correctness of the sync - if that is not the case here, it's less of a problem than I initially assumed. Thanks.

cammellos replied:
I don't think that would be possible, essentially non-repudiation and integrity still stand, and the result is always unambiguous (we never sign the parsed message but always the byte-stream), even if there's  more than one way to serialize the message (due to the order of fields not being specified/enforced essentially, given `protobuf` is map).

There can be no doubt on the origin and the integrity of the message, whether there might be some doubt on the interpretation it would be a problem in the protocol itself (say for example the two clients have mismatched protobuf specifications).
In general protobuf being not self-describing might lead to such issues, but they would come up regardless of the signature scheme used.

thanks for the feedback

oskarth replied:
This question seems to come up quite frequently, it might be a good idea to add it to a rationale/FAQ section at the end so we don't end up having the same conversation over and over again :)

arnetheduck commented on status-payloads-spec.md:
```diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
-| 3 | name | `string` | The self-assigned name of the device |
-| 4 | fcm-token | `string` | The FCM Token used by mobile devices for push notifications |
+| 1 | clock | `uint64` | clock value of the chat | 
+| 2| installation_id | `string` | A randomly generated id that identifies this device |
```
```suggestion
| 2 | installation_id | `string` | A randomly generated id that identifies this device |
```

arnetheduck commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+    MEMBERS_ADDED = 3;
+    MEMBER_JOINED = 4;
+    MEMBER_REMOVED = 5;
+    ADMINS_ADDED = 6;
```
how does the protocol establish a total order of these events?

cammellos replied:
We use clock value (lamport timestamp essentially), ties broken by message-id , same as chat messages essentially

cammellos commented:
@decanus updated according the feedback, could you take another look please?

decanus commented:
@cammellos I need to reconsider the signature part, will get back to you asap.

decanus commented:
Also, @cammellos protobuf version should be specified somewhere.

adambabik commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
-Each membership update MUST be sent with a corresponding `chat-id`. 
-The format of this chat id MUST be a string, [UUID](https://tools.ietf.org/html/rfc4122 ), concatenated with the hex-encoded public key of the creator of the chat. This chat-id MUST be validated by all clients, and MUST be discarded if it does not follow these rules.
+Each membership update MUST be sent with a corresponding `chatId`. 
+The format of this chat id MUST be a string, [UUID](https://tools.ietf.org/html/rfc4122 ), concatenated with the hex-encoded public key of the creator of the chat. This chatId MUST be validated by all clients, and MUST be discarded if it does not follow these rules.
```
```suggestion
The format of this chat ID MUST be a string of [UUID](https://tools.ietf.org/html/rfc4122) concatenated with the hex-encoded public key of the creator of the chat. This chat ID MUST be validated by all clients, and MUST be discarded if it does not follow these rules.
```

Do we use `-` for concatenation or nothing?

cammellos replied:
we do, updated

adambabik commented on status-payloads-spec.md:
```diff
@@ -114,15 +168,19 @@ Message types are required to decide how a particular message is encrypted and w
 `timestamp` MUST be Unix time calculated when the message is created in milliseconds. This field SHOULD not be relied upon for message ordering.
 
-`clock` SHOULD be calculated using the algorithm of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps). When there are messages available in a chat, `clock`'s value is calculated based on the last received message in a particular chat: `last-message-clock-value + 1`. If there are no messages, `clock` is initialized with `timestamp * 100`'s value.
+`clock` SHOULD be calculated using the algorithm of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps). When there are messages available in a chat, `clock`'s value is calculated based on the last received message in a particular chat: `max(timeNowInMs, last-message-clock-value + 1)`. If there are no messages, `clock` is initialized with `timestamp * 100`'s value.
```
Why `* 100`?

cammellos replied:
that's actually not the case anymore, we multiplied by `100` to separate the timestamp from the  `sequence` number of the message, but it's not really needed, so we just use a timestamp now, updated

oskarth commented on status-group-chats-spec.md:
````diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+
+The protobuf messages are:
+
+```protobuf
````
1. Could we specify protobuf version used in these fragments? Either globally or per section

2. If we have a package name, can we specify these? https://developers.google.com/protocol-buffers/docs/proto If we don't, could we consider adding it?

cammellos replied:
I will add the version, currently we use `protobuf` as a package name (in go it does not have change the generated code, so we can pick anything), what should we go for? (I might update the specs though in a separate PR, once the change is made)

decanus replied:
Yeah we should defo specify, in MVDS we also specify which version we use.

decanus replied:
@cammellos @oskarth you can add `syntax = "proto3";` to the protocol buffer itself as the first line before the message.

cammellos replied:
just to clarify, we do always specify both `syntax` (`proto3`) and package (`protobuf`, the package name we use is just `protobuf`), just wondering whether we should use a more meaningful package name

oskarth replied:
What about `status.messaging`? Looking at https://github.com/status-im/status-go/tree/baa0767c263fa25fdd8d90a42d13f1f1fffce804/protocol/protobuf it seems like there are 3 short-ish protobufs, so having one namespace for all of them seems good enough for them. Once we do different domains, we can have `status.foo`, and we'll also have `vac.waku` etc.

oskarth commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+message MembershipUpdateMessage {
+  // The chat id of the private group chat
+  string chat_id = 1;
+  // A list of events for this group chat, first x bytes are the signature, then is a 
```
What is x?

cammellos replied:
65 bytes, added

oskarth commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+// MembershipUpdateMessage is a message used to propagate information
+// about group membership changes.
+message MembershipUpdateMessage {
+  // The chat id of the private group chat
```
Is this unique to private group chats? vs "group chat" as a construct (I know we don't - currently, anymore - support public group chats).

cammellos replied:
Yes, that's only used for private group chats

oskarth commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
-  "signature": string,
-  "chat-id": string
+message MembershipUpdateEvent {
+  // Lamport timestamp of the event
```
can we put a reference to where in the specs this is elaborated on?

oskarth commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+message MembershipUpdateEvent {
+  // Lamport timestamp of the event
+  uint64 clock = 1;
+  // List of public keys of objects of the action
```
can we rephrase this? "of objects" seems unclear to me

cammellos replied:
the best I could come up with is `of the targets`, not sure is much clearer though :)

oskarth commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+  // The type of the event
+  EventType type = 4;
+
+  enum EventType {
```
a reference to semantics and guarantees in different part of spec would be useful

oskarth commented on status-payloads-spec.md:
```diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
 ## Encoding
 
-The payload is encoded using [Transit format](https://github.com/cognitect/transit-format). Transit was chosen over JSON in order to reduce the bandwidth.
+The payload is encoded using [Protobuf](https://developers.google.com/protocol-buffers)
```
specify version

oskarth commented on status-payloads-spec.md:
```diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
 ### Message
 
-The type `Message` represents a text message exchanged between clients and is identified by the transit tag `c4`.
+The type `ChatMessage` represents a text message exchanged between clients.
```
text -> chat? text to me seems to imply plain text

oskarth commented on status-payloads-spec.md:
```diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
+message ChatMessage {
+  // Lamport timestamp of the chat message
+  uint64 clock = 1;
+  // Unix timestamps in milliseconds, currently not used as we use whisper as more reliable, but here
```
I think I know what you mean, but this can probably be rephrased for clarity

cammellos replied:
Not sure it's actually even worth mentioning is not used, the only way we use it is  cosmetic (to show a user what time a message has been sent, but it's purely cosmetic (another client might chose not to do so), what do you think?

oskarth commented on status-payloads-spec.md:
```diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
+| 1 | clock | `uint64` | The clock of the chat|
+| 2 | timestamp | `uint64` | The sender timestamp at message creation |
+| 3 | text | `string` | The content of the message |
+| 4 | response_to | `string` | The ID of the message replied to |
```
would it make sense to have an ID type with restrictions to it? even if it is something like `stringid` and then elaborated on above

cammellos replied:
I would probably keep it a string, because technically it is a string (so strings need to be used in protobuf), and then is validated (in the same way for other fields), I have always taken this as what is currently sent across the wire, but never really spent time questioning it.

Essentially, do we want this field to represent what is being sent in `protobuf` or we want it to diverge from that in the specs so that it's easier to interpret for someone reading it?

I don't have a strong opinion myself, either are fine

oskarth replied:
Keeping it as a string here makes sense for accuracy

One idea would be to have a small section below on restriction/validation to explain additional requirements we have

oskarth commented on status-payloads-spec.md:
```diff
@@ -65,46 +65,100 @@ If a signature is not present but an author is provided by a layer below, the me
 
 The following content types MUST be supported:
-* `text/plain` identifies a message which content is a plain text.
+* `TEXT_PLAIN` identifies a message which content is a plain text.
```
```suggestion
* `TEXT_PLAIN` identifies a message which content is plaintext.
```

oskarth commented on status-payloads-spec.md:
````diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
 The details are in the  [Group chats specs](status-group-chats-spec.md)
 
+```protobuf
+message MembershipUpdateEvent {
````
I feel like I have seen this before, is there a reason to duplicate it?

cammellos replied:
Moved to `status-group-chats-spec.md`

oskarth commented on status-payloads-spec.md:
```diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
-- Struct fields can be enriched with a new key, which will be ignored by old clients.
-- An element can be appended to the `Transit` array, which will also be ignored by old clients.
+- Accretion is always supported
+- Deletion of existing is not supported and might break compatibility
```
```suggestion
- Deletion of existing fields or messages is not supported and might break compatibility
```

oskarth commented on status-payloads-spec.md:
```diff
@@ -134,94 +192,169 @@ All incoming messages can be matched against a chat. Below you can find a table
 
 TBD.
 
 ## Design rationale
```
so now we don't even have a rationale! pff ;)

oskarth reviewed (changes requested):
Awesome work on moving everything over to protobuf!

Initial comments

cammellos commented:
@oskarth addressed feedback and left a few questions, I have specified version of `protobuf` in `status-payload-specs.md` (saying that is `proto3` unless specified otherwise), as we use `protobuf` throughout the whole stack

decanus commented on status-client-spec.md:
```diff
@@ -67,6 +67,10 @@ Implementing a Status clients means implementing the following layers. Additiona
 
+### Protobuf
+
+We use `protobuf` in different layers, the version used is `proto3` unless stated otherwise.
```
```suggestion
We use [`protobuf`](https://developers.google.com/protocol-buffers/) in different layers, the version used is `proto3` unless stated otherwise.
```

decanus commented on status-group-chats-spec.md:
```diff
@@ -28,177 +28,123 @@ This documents describes the group chat protocol used by the status application.
+message MembershipUpdateMessage {
+  // The chat id of the private group chat
+  string chat_id = 1;
+  // A list of events for this group chat, first 65 bytes are the signature, then is a 
```
why do this and not just create a new internal message called event

```
message Event {
 repeated bytes signature = 1;
 MembershipUpdateEvent update = 2;
}
```

Not criticial but worth knowing the rationale.

cammellos replied:
We sign the raw bytes, so those need to be sent over and easily extracted.
So your suggestion would be:

```
message Event {
 repeated bytes signature = 1;
 repeated bytes payload = 2; // This is actually a MembershipUpdateEvent
}
```

which works fine, and it is more explicit.

I slightly prefer the concatenation mainly for ease of use, clients will take the event and store it as it, unchanged, so it can be relayed to third party, and probably save a few bytes.

Either though are perfectly fine I think.

decanus replied:
@cammellos ok that makes sense, lets keep it as is then.

decanus commented on status-payloads-spec.md:
```diff
@@ -114,15 +167,19 @@ Message types are required to decide how a particular message is encrypted and w
+
+Messages with a `clock` greater than `120` seconds over the whisper timestamp SHOULD be discarded, in order to avoid malicious users to increase the `clock` of a chat arbitrarily.
+
+Messages with a `clock` less than `120` might indicate an attempt to insert messages in the chat history, this is not distinguishable though from a re-transmit from the `datasync` layer. Client MIGHT mark this messages with a warning to the user, or discard them.
```
```suggestion
Messages with a `clock` less than `120` might indicate an attempt to insert messages in the chat history, this is not distinguishable though from a re-transmit from the `datasync` layer. A client MAY mark this messages with a warning to the user, or discard them.
```

MIGHT is not in rfc2119

cammellos committed:
Apply suggestions from code review

Co-Authored-By: Dean Eigenmann <dean.eigenmann@icloud.com>

oskarth reviewed (approved):
Thanks! Left a comment re namespace and some other stuff, otherwise lgtm https://github.com/status-im/specs/pull/80/files#r369364189
